Modification of the entry point to viewController instead of MainStoryBoard

ronaldngounou opened the pull request:
The old version was starting the app from the main story board and this version is starting the app from the view Controller

ronaldngounou committed:
change entry point to start from viewController instead of mainStoryBoard. The old version was starting the app from the main story board and this version is starting the app from the view Controller

aleksandrshoshiashvili commented on Quizzler-iOS13/SceneDelegate.swift:
```diff
@@ -13,11 +13,34 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+        let window = UIWindow(windowScene: windowScene)
+        
+        // 3. Create the view hierarchie programmatically
+        let viewController = ViewController()
```
Does it actually work? Given that you have your UI inside Storyboard, you should initialize your ViewController from storyboard. 

`let viewController = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: "ViewController")`

This might give you a hint:
https://medium.com/@vasanthkumarvelusamy/programmatic-entry-to-viewcontroller-in-storyboard-on-app-launch-51c3281ceea2

ronaldngounou replied:
I followed the steps in the Medium blog post and updated my files on a recent commit. This error is showing

This error appears on line 21 of the AppDelegate file: "Cannot convert value of type 'UIViewController.Type' to expected argument type 'UIViewController'" 

Do you have any idea on this?

aleksandrshoshiashvili commented on Quizzler-iOS13/SceneDelegate.swift:
```diff
@@ -13,11 +13,34 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+        
+        // 3. Create the view hierarchie programmatically
+        let viewController = ViewController()
+        let navigationController = UINavigationController(rootViewController: viewController)
```
_Side note, not a scope of this PR_

For a single page application it is a valid solution, at the same time the more pages and the more conditions to when open which page you'll have — the more difficult it be to handle navigation in this way.

There are different techniques how to improve it, one way is to use Coordinator pattern:
https://medium.com/@maysam.shahsavari/a-comprehensive-guide-to-coordinator-pattern-in-swift-7e7647ecc525
https://medium.com/@dkw5877/flow-coordinators-333ed64f3dd
https://www.hackingwithswift.com/articles/71/how-to-use-the-coordinator-pattern-in-ios-apps

ronaldngounou replied:
Thank you! I see 👍🏽

ronaldngounou committed:
Add entry to ViewController in Storyboard on App launch

ronaldngounou committed:
Add entry to ViewController instead of MainStoryBoard

aleksandrshoshiashvili commented on Quizzler-iOS13/AppDelegate.swift:
```diff
@@ -10,31 +10,32 @@ import UIKit
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        
+        let storyboard = UIStoryboard(name: "Main", bundle: nil)
+        let viewController = storyboard.instantiateViewController(identifier: "ViewController") as! ViewController
```
As we've discussed, we should avoid to use force unwrap
In this specific case it should be easy, as
- storyboard.instantiateViewController returns non-optional UIViewController
- we cast here to ViewController type, but we actually do not use it ViewController properties/functions anyhow

So you can just drop "as! ViewController" part and it should also work

ronaldngounou replied:
In the commit "[8ada50a](https://github.com/ronaldngounou/Quizzler-iOS13/pull/2/commits/8ada50a432ddafffaecd5b940f196579bc7cc1e1)" I deleted force unwrap. The old version was not launching. In this new version, the code is running but the applications stays on a black screen. The error says: "Failed to instantiate the default view controller for UIMainStoryboardFile 'Main' - perhaps the designated entry point is not set?"

To solve this issue, I tried added the 'Main' value for the StoryBoard property in the Info.plist.

Still not working.

aleksandrshoshiashvili commented on Quizzler-iOS13/AppDelegate.swift:
```diff
@@ -10,31 +10,32 @@ import UIKit
 
+    
+    
+    // This function should stay?
```
You can delete for now and add only if you'll need it (if you going to do something on "app will terminate" event)

ronaldngounou replied:
Thank you!

ronaldngounou committed:
Delete force unwrap. The old version was not launching. In this new version, the code is running but it stays on a black screen. The error says: "Failed to instantiate the default view controller for UIMainStoryboardFile 'Main' - perhaps the designated entry point is not set?"
To solve this issue, I tried added the 'Main' value for the StoryBoard property in the Info.plist.

ronaldngounou commented:
In the commit "[8ada50a](https://github.com/ronaldngounou/Quizzler-iOS13/pull/2/commits/8ada50a432ddafffaecd5b940f196579bc7cc1e1)"  I deleted force unwrap. The old version was not launching. In this new version, the code is running but the applications stays on a black screen. The error says: "Failed to instantiate the default view controller for UIMainStoryboardFile 'Main' - perhaps the designated entry point is not set?"

To solve this issue, I tried added the 'Main' value for the StoryBoard property in the Info.plist.

Still not working.

ronaldngounou committed:
Modify usage of force unwrap

Modified the usage of force unwrap. The old version was using force unwrapping which is a sign of bad design. The new version handles force unwrapping by raising an error in case of failure.

ronaldngounou committed:
Remove mutating and public classes

Mutating in a structure is often a sign of bad design. The old version was using mutatings inside the struct. The current version uses a final class QuizBrain with public functions inside the QuizBrain model.

aleksandrshoshiashvili commented on Quizzler-iOS13/Controller/ViewController.swift:
```diff
@@ -8,7 +8,7 @@
 import UIKit
 
-class ViewController: UIViewController {
+public class ViewController: UIViewController {
```
Why do you need public? You going to use ViewController inside the same module, so should be internal (default)

ronaldngounou replied:
Alright. I thought the default class was public. 

I still have the same issue on "Failed to instantiate the default view controller for UIMainStoryboardFile 'Main' - perhaps the designated entry point is not set?" and the app stays on a black screen.

ronaldngounou replied:
<img width="951" alt="Screenshot 2024-04-15 at 1 08 48 AM" src="https://github.com/ronaldngounou/Quizzler-iOS13/assets/74538524/1378034c-5413-4130-8f0f-4638b2a47cb6">

I saw your comment on removing references to Main.storyboard in the info.plist. I deleted the two lines underlined in red. 
I also tried to change the "LaunchScreen interface base name" to ViewController but the screen remains black. 
I also read and applied the suggestions proposed [here](https://stackoverflow.com/questions/20875823/failed-to-instantiate-the-default-view-controller-for-uimainstoryboardfile-main) but I am still stucked.


Thank you!

aleksandrshoshiashvili replied:
There are 2 ways to fix it:

1. Restore SceneDelegate and then initialize window like this
`func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {

    if let windowScene = scene as? UIWindowScene {
        let window = UIWindow(windowScene: windowScene)
        // window.rootViewController = ...
        self.window = window
        window.makeKeyAndVisible()
    }
}`

2. Get rid of SceneDelegate approach, for that you need to also remove Application Scene Manifest from Info.plist

1st approach is preferable though

> I also tried to change the "LaunchScreen interface base name" to ViewController but the screen remains black.

Keep it as it was :)

ronaldngounou replied:
I followed the first approach. I had to copy and paste the old SceneDelegate file from a previous state and create a new file. I searched how to restrore a project from a state but I was not able to figure out how to show the terminal in xcode. 

In my latest commit [94b81cf](https://github.com/ronaldngounou/Quizzler-iOS13/pull/2/commits/94b81cfd52ab2e91bbfce49128083cffdbb53a0b), I have some issues regarding nil values.

ronaldngounou committed:
Remove public class for ViewController

ronaldngounou committed:
Error with the screen launch stopped occuring but blocked with unwrapping

In this version, I am stucked with 2 errors stating "unexpectedly found nil while implicitly unwrapping an optional value" I tried to resolve it by following the same logic but still not working.

ronaldngounou committed:
Error with the screen launch stopped occuring but blocked with unwrapping

In this version, I am stucked with 2 errors stating "unexpectedly found nil while implicitly unwrapping an optional value" I tried to resolve it by following the same logic but still not working.

ronaldngounou commented:
I have no clue why this commit is empty.. [94b81cf](https://github.com/ronaldngounou/Quizzler-iOS13/pull/2/commits/94b81cfd52ab2e91bbfce49128083cffdbb53a0b)

ronaldngounou committed:
Force unwrapping error

Unexpectedly found nil while implicitly unwrapping an Optional value

ronaldngounou committed:
Merge remote-tracking branch 'refs/remotes/origin/Version1'

Conflicts:
	Quizzler-iOS13/SceneDelegate.swift

ronaldngounou committed:
working version with a start from ViewController

ronaldngounou commented:
Working version with a start from ViewController
